ariadne_codegen/plugins/shorter_results.py

```diff
@@ -0,0 +1,179 @@
+    me = (await client.get_authenticated_user).me
+
+By enabling this plugin the return type of the generated
+`get_authenticated_user` would instead directly return a
+`GetAuthenticatedUserMe` so the caller could use:
+
+    me = await client.get_authenticated_user
+
+This plugin can be enabled by either adding the plugin in the settings:
+
+    plugins = ["ariadne_codegen.ShorterResultsPlugin"]
+
+Or with the shorthand setting:
+
+    [tool.ariadne-codegen]
+    unwrap_single_fields = true
+"""
+
+import ast
+from typing import Union
+
+from ariadne_codegen.plugins.base import (
+    Plugin,
+    SelectionSetNode,
+    OperationDefinitionNode,
+)
+
+
+class ShorterResultsPlugin(Plugin):
+    """
+    Make single field return types expanded.
+
+    All client method that returns type with a single field will be expanded and
+    instead return the type of the inner field.
+    """
+
+    def generate_result_class(
+        self,
+        class_def: ast.ClassDef,
+        operation_definition: OperationDefinitionNode,
+        selection_set: SelectionSetNode,
+    ) -> ast.ClassDef:
+        """Store a map of all classes and its ast"""
+        self.class_dict[class_def.name] = class_def
+
+        return super().generate_result_class(
+            class_def, operation_definition, selection_set
+        )
+
+    def generate_client_module(self, module: ast.Module) -> ast.Module:
+        """Add extra import needed after expanding the inner field"""
+        if len(self.extended_imports) == 0:
+            return super().generate_client_module(module)
+
+        for stmt in module.body:
+            # We know we've already imported the wrapping type so we're only
+            # interested in `ImportFrom`.
+            if not isinstance(stmt, ast.ImportFrom):
+                continue
+
+            # The module we import from is always the same as the method we
+            # modified when changing the return type.
+            if stmt.module not in self.extended_imports:
+                continue
+
+            # Add all additional imports discovered when generating the client
+            # method.
+            for additional_import in self.extended_imports[stmt.module]:
+                stmt.names.append(ast.alias(name=additional_import))
+
+        return super().generate_client_module(module)
+
+    def generate_client_method(
+        self, method_def: Union[ast.FunctionDef, ast.AsyncFunctionDef]
+    ) -> Union[ast.FunctionDef, ast.AsyncFunctionDef]:
+        """
+        Change the method generated in the client to call the inner field and
+        change the return type.
+        """
+        if len(method_def.body) < 1:
+            return super().generate_client_method(method_def)
+
+        # TODO: Can we assume that the return statement is always the last
+        # statement?
+        return_stmt = method_def.body[-1]
+        if not isinstance(return_stmt, ast.Return):
+            return super().generate_client_method(method_def)
```

Comment:
This will obviously break if the codegen introduces a new path for returns but right now it seems a bit unlikely. I guess the alternative would be to store all return statements in a list, filter out those that's not for the generated class (f.ex. if there are `None` for optionals) and then update all.

I didn't do that because such refactor could change this as well instead of doing pre-mature optimization but maybe there's a need for it if someone else chains plugins and change the ast with a custom plugin introducing this?

Reply:
Cool! Kept as is because I want to check lengt of generator and see if we're a return or async for statement!